Introduce PhysicalServerProvision StateMachine

miha-plesko opened the pull request:
With this commit we implement a new external state machine
    
```
PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision
```
    
which in essence invokes (also new) internal state machine and then polls for    its status. The external state machine uses AE messages to allow for different     customization for redfish-managed servers than for lenovo-manager ones.

Depends on 
https://github.com/ManageIQ/manageiq/pull/18573
https://github.com/ManageIQ/manageiq-automation_engine/pull/306

@miq-bot add_label enhancement

lfu commented on content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Lifecycle.class/provisioning.yaml:
```diff
@@ -0,0 +1,16 @@
+    description: 
+  fields:
+  - Relationship4:
+      value: "/ManageIQ/System/Request/InspectMe"
```
This is useful for debugging. But should not be in production code.

lfu commented on content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Lifecycle.class/provisioning.yaml:
```diff
@@ -0,0 +1,16 @@
+  - Relationship4:
+      value: "/ManageIQ/System/Request/InspectMe"
+  - Relationship5:
+      value: "/Physical_Infrastructure/PhysicalServer/Provisioning/Profile/${/#user.normalized_ldap_group}#get_state_machine"
```
All ```Physical_Infrastructure``` should be changed to ```PhysicalInfrastructure```.

lfu commented on content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/common_provision.rb:
```diff
@@ -0,0 +1,17 @@
+$evm.log(:info, "options = #{options}")
+$evm.log(:info, "pxe_server_id = #{options[:pxe_server_id]}")
+$evm.log(:info, "src_pxe_image_id = #{options[:src_pxe_image_id]}")
+$evm.log(:info, "src_configuration_profile_id = #{options[:src_configuration_profile_id]}")
```
For the newer methods we are switching to a new format which helps us to test the methods easily.

For example you can look at https://github.com/ManageIQ/manageiq-content/blob/2b6360cc885c7f099838c84c121b8fe750ffb8ba/content/automate/ManageIQ/Cloud/VM/Provisioning/Placement.class/__methods__/best_fit_amazon.rb

lfu commented on content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/redfish_provision.rb:
```diff
@@ -0,0 +1,26 @@
+rescue => err  
+  $evm.log(:error, "[#{err}]\n#{err.backtrace.join("\n")}")
+  $evm.root['ae_result'] = 'error'
+  exit MIQ_STOP
```
Automate does the error handling for these methods.

lfu commented on content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Provisioning/StateMachines/PhysicalServerProvision.class/__methods__/update_provision_status.rb:
```diff
@@ -0,0 +1,39 @@
+  $evm.create_notification(:level => "error", :subject => prov.miq_request, \
+                           :message => "PhysicalServer Provision Error: #{updated_message}")
+  $evm.log(:error, "PhysicalServer Provision Error: #{updated_message}")
+end
```
Please convert to new format.

lfu commented on content/automate/ManageIQ/Physical_Infrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/default_provision.rb:
```diff
@@ -0,0 +1,4 @@
+#
+# Description: <Method description here>
+#
+$evm.log(:info, "BUREK PROVISIONING DEFAULT PHYSICAL SERVER")
```
Same here for new format.

lfu commented on content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/PhysicalServerProvision.class/__methods__/update_provision_status.rb:
```diff
@@ -0,0 +1,65 @@
+                  @handle = handle
+                end
+
+                def main
```
Please break this into small methods so each one may be around 6~7 lines.

lfu commented on content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/PhysicalServerProvision.class/__methods__/update_provision_status.rb:
```diff
@@ -0,0 +1,65 @@
+end
+
+ManageIQ::Automate::PhysicalInfrastructure::PhysicalServer::Provisioning::StateMachines::PhysicalServerProvision::UpdateProvisionStatus.new.main
+
```
Please remove the empty lines.

miha-plesko commented:
Thanks for reviewing @lfu , I've pushed an updated version with your suggestions applied plus unit tests for all the new methods.

BTW I'm wondering whether we need to change anything in order to support scenario where single Request would be realized with more than one Task? If I create request with, say, three source ids

```
src_id = [1, 2, 3]
```

then the three servers will get provisioned, but there seems to be "race condition" between them as they are all updating same request's status. Is there perhaps any example on how to handle such scenario?

miha-plesko commented:
Marking as WIP again, because we need to merge core changes first or else Travis goes crazy

lfu commented on content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/check_provision.rb:
```diff
@@ -0,0 +1,45 @@
+                  case result
+                  when 'error'
+                    @handle.root['ae_result'] = 'error'
+                    @handle.root['ae_reason'] = task.message.to_s.sub('Error: ', '')
```
Why do we need ```to_s```? ```task.message``` returns a string.

miha-plesko replied:
Good point

lfu commented on content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status.rb:
```diff
@@ -0,0 +1,71 @@
+                      exit(MIQ_STOP)
+                    end
+                    task
+                  end
```
```
@task ||= @handle.root['miq_provision_task'].tap do |task|
  unless task
    @handle.log(:error, 'miq_provision_task object not provided')
    exit(MIQ_STOP)
  end
  unless task.source
    @handle.log(:error, 'miq_provision_task.source object not provided (PhysicalServer expected)')
    exit(MIQ_STOP)
  end
end
```

miha-plesko replied:
You know, it felt like there is a better way, thanks 👍

lfu commented on spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods.class/__methods__/provision_spec.rb:
```diff
@@ -0,0 +1,17 @@
+  let(:root_object) { Spec::Support::MiqAeMockObject.new(:miq_provision_task => svc_task) }
+  let(:ae_service)  { Spec::Support::MiqAeMockService.new(root_object) }
+
+  it 'execute is called on task' do
```
```
it 'calls execute on task' do
```

miha-plesko replied:
Done

lfu commented on spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb:
```diff
@@ -0,0 +1,68 @@
+    before { ae_service.inputs['status'] = 'the status' }
+    it do
+      expect(subject.status_param).to eq('the status')
+    end
```
```
it { expect(subject.status_param).to eq('the status') }
```

lfu commented on spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb:
```diff
@@ -0,0 +1,68 @@
+      root_object['ae_state'] = 'at state'
+    end
+    context 'updates task and request message' do
+      it do
```
Please add a description for ```it```.

miha-plesko replied:
Done

lfu commented on spec/content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/__methods__/update_provision_status_spec.rb:
```diff
@@ -0,0 +1,68 @@
+
+    context 'fires notification on error' do
+      before { root_object['ae_result'] = 'error' }
+      it do
```
Same here.

miha-plesko replied:
Done

miha-plesko commented:
Sorry @lfu had to change it a little bit: I've replaced all occurances of

```
$evm.root['miq_provision_task']
```

with

```
$evm.root['physical_server_provision_task']
```

because we've modified automation engine a bit, see here https://github.com/ManageIQ/manageiq/pull/18573/files#r277710290

lfu commented on content/automate/ManageIQ/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Provision.class/default.yaml:
```diff
@@ -0,0 +1,36 @@
+    description: 
+  fields:
+  - PreProvision:
+      value: "/PhysicalInfrastructure/PhysicalServer/Provisioning/StateMachines/Methods/PreProvision#${/#miq_provision_task.source.ems_type}"
```
Is ```miq_provision_task``` used in this file by purpose?

miha-plesko replied:
Thanks

lfu commented:
LGTM. Please remove WIP from the title when PR is ready to merge.

miha-plesko commented:
@lfu slightly updated again following @gmcculloug 's advice here https://github.com/ManageIQ/manageiq/pull/18573#discussion_r279118400. This time I replaced

```
source.ems_type -> source.emstype
```

lfu commented:
LGTM 👍 Remember to remove WIP from the title when PR is ready to merge.
